src/main/kotlin/com/avail/utility/structures/EnumMap.kt

```diff
@@ -0,0 +1,210 @@
+	 * Answer the value at the provided key. If no value is available (`null`
+	 * is at [Enum.ordinal] in the [sourceValues] array), evaluate the lambda,
+	 * place the resulting value in this map at that key, then answer the value.
+	 *
+	 * @param key
+	 *   The key to look up.
+	 * @param mappingFunction
+	 *   The function to run to produce a value if none found.
+	 * @return The stored value or the result of the [mappingFunction].
+	 */
+	fun getOrPut (key: K, mappingFunction: (K) -> V): V
+	{
+		var value = sourceValues[key.ordinal]
+		if (value == null)
+		{
+			value = mappingFunction(key)
+			sourceValues[key.ordinal] = value
+		}
+		return value
+	}
+
+	/**
+	 * Removes the specified key and its corresponding value from this
+	 * [EnumMap].
+	 *
+	 * @return
+	 *   The previous value associated with the key, or `null` if the key was
+	 *   not present in the map.
+	 */
+	fun remove(key: K): V?
+	{
+		val value = sourceValues[key.ordinal]
+		sourceValues[key.ordinal] = null
+		return value
+	}
+
+	/**
+	 * Removes all elements from this [EnumMap].
+	 */
+	fun clear()
+	{
+		for (i in sourceValues.indices)
+		{
+			sourceValues[i] = null
+		}
+	}
+
+	override fun containsKey(key: K): Boolean = true
+
+	override fun containsValue(value: V): Boolean =
+		sourceValues.any { value == it }
+
+	override fun get(key: K): V = sourceValues[key.ordinal]!!
+
+	override fun isEmpty(): Boolean = false
+
+	override val entries: Set<Map.Entry<K, V>>
+		get() = enums.map {
+			object : Map.Entry<K, V>
+			{
+				override val key: K = it
+				override val value: V = sourceValues[it.ordinal]!!
+			}
+		}.toSet()
+
+	/**
+	 * Construct an [EnumMap].
+	 *
+	 * @param enums
+	 *   The [Array] of [Enum]s, in [Enum.ordinal] order of all the `Enum`
+	 *   values represented in the `Enum`. It is expected that **all** enum
+	 *   values will be present.
+	 */
+	constructor (enums: Array<K>) :
+		this(enums, Array<Any?>(enums.size) { null } as Array<V?>)
+
+	companion object
+	{
+		/**
+		 * Answer a new [EnumMap].
+		 *
+		 * @param enums
+		 *   The [Array] of [Enum]s, in [Enum.ordinal] order of all the `Enum`
+		 *   values represented in the `Enum`. It is expected that **all** enum
+		 *   values will be present.
+		 * @param values
+		 *   The vararg [Array] of values stored in this `Enum`.
+		 */
+		fun <K : Enum<K>, V : Any> enumMap (enums: Array<K>, vararg values: V)
+			: EnumMap<K, V> =
+				EnumMap(enums, Array<Any?>(enums.size) { values[it] }
+					as Array<V?>)
+
+		/**
+		 * Answer a new [EnumMap].
+		 *
+		 * @param enums
+		 *   The [Array] of [Enum]s, in [Enum.ordinal] order of all the `Enum`
+		 *   values represented in the `Enum`. It is expected that **all** enum
+		 *   values will be present.
+		 * @param populator
+		 *   A lambda that accepts a value of the Enum and answers a value to be
+		 *   stored at the enum value key.
+		 */
+		fun <K : Enum<K>, V : Any> enumMap (
+			enums: Array<K>,
+			populator: (K) -> V?): EnumMap<K, V> = EnumMap(
```

Comment:
Too much is happening on this line.  It's the final argument, the result type, and the start of an invocation of a constructor.  Breaking it just before the ")" breaks this up in such a way that there are clearly two arguments, there is definitely a return type and expression body.  If the expression body had a more complex beginning, I would also move it to yet another line.